zebra/rt_netlink.c

```diff
@@ -1767,6 +1767,33 @@ ssize_t netlink_route_multipath_msg_encode(int cmd,
 		nl_attr_nest_end(&req->n, nest);
 	}
 
+	/*
+	 * Always install blackhole routes without using nexthops, because of
+	 * the following kernel problems:
+	 * 1. Kernel nexthops don't suport unreachable/prohibit route types.
+	 * 2. Blackhole kernel nexthops are deleted when loopback is down.
+	 */
+	for (ALL_NEXTHOPS_PTR(dplane_ctx_get_ng(ctx), nexthop)) {
```

Comment:
So for blackhops, this should only ever be one nexthop or one nexthop recursively resolved to a blackhole nexthop.

I don't think we should take the perf hit of going over EVERY nexthop to check for every route install blackhole or not. 

I would just manually check the nexthop and ->resolved pointer here to make it clear what you are doing rather than a loop.

Reply:
this should be prohibited in staticd not zebra

Reply:
@donaldsharp you're right, and I have a plan to fix this in staticd, but this has nothing to do with this PR

Reply:
it's fine for zebra to enforce some consistency, even if a daemon sends something ambiguous - it's fine for both staticd and zebra to do some validation and enforcement imo.

Reply:
i disagree it's not another issue, you have changed the code such that zebra has to loop over every nexthop (including recursives) before every route install to check for a blackhole when 99% of the time it wont be there. The code before only did this if it was ecmp=1.

If it was necessary to check all nexthops sure... but we know at least for blackholes its not so why take the perf hit?

Reply:
> If it was necessary to check all nexthops sure... but we know at least for blackholes its not so why take the perf hit?

I don't think there's a real perf hit to looking through this list, really. we're not operating at vpp-like scale here, and looking at a few pointers isn't going to make a meaningful difference (especially compared to the system-call to actually do the netlink programming)?

Reply:
I agree with Mark.

First, there is already a loop over all nexthops a little bit later in the code to calculate the number of nexthops. Nobody complains about the performance hit there.
Second, there's currently no mechanism to ensure that blackhole nexthop can't be combined with other nexthops. If any other daemon can send such route to zebra, then zebra must check for this.

Reply:
If we are checking for mixed nexthops with one being a blackhole.  deep in the rt_netlink.c code is not the right place to do any such thing.  It should be done at a higher level in zebra so that all dataplanes can take advantage of it.

Reply:
@sworleys @donaldsharp I removed the loop, please check it out.